wapitiCore/attack/mod_csrf.py

```diff
@@ -0,0 +1,203 @@
+#!/usr/bin/env python3
+from collections import Counter
+import math
+from requests.exceptions import ReadTimeout
+
+from wapitiCore.attack.attack import Attack
+from wapitiCore.language.vulnerability import Vulnerability, Anomaly, _
+from wapitiCore.net.web import Request
+from wapitiCore.net.crawler import Page
+
+class mod_csrf(Attack):
+    """
+    This class implements a CSRF/XSRF attack
+    """
+
+    name = "csrf"
+
+    do_get = True
+    do_post = True
+
+    csrf_string = None
+
+    MIN_ENTROPY = 3
+    MIN_LENGTH = 8
+
+    TOKEN_FORM_STRINGS = [
+        "authenticity_token", "_token", "csrf_token", "csrfname", "csrftoken", "anticsrf",
+        "__requestverificationtoken", "token", "csrf", "_csrf_token", "xsrf_token",
+        "_csrf", "csrf-token", "xsrf-token", "_wpnonce", "fkey"
+        ]
+
+    TOKEN_HEADER_STRINGS = [
+        "csrf-token", "x-csrf-token", "xsrf-token", "x-xsrf-token", "csrfp-token",
+        "anti-csrf-token", "x-csrf-header", "x-xsrf-header", "x-csrf-protection"
+    ]
+
+    @staticmethod
+    def entropy(string: str):
+        # Shannon entropy calculation
+        # https://stackoverflow.com/questions/15450192/fastest-way-to-compute-entropy-in-python
+        probabilities = [n_x / len(string) for x, n_x in Counter(string).items()]
+        e_x = [-p_x * math.log(p_x, 2) for p_x in probabilities]
+        return sum(e_x)
+
+    def is_csrf_present(self, original_request: Request):
+        """Check whether anti-csrf token is present"""
+
+        # Look for anti-csrf token in form params
+        for param in original_request.post_params:
+            if param[0].lower() in self.TOKEN_FORM_STRINGS:
+                self.csrf_string = param[0]
+                return param[1]
+
+        # Look for anti-csrf token in HTTP headers
+        for header in original_request.headers:
+            if header.lower() in self.TOKEN_HEADER_STRINGS:
+                self.csrf_string = header
+                return original_request.headers[header]
+
+        return None
+
+    def is_csrf_robust(self, csrf_value):
+        """Check whether anti-csrf token is robust enough"""
+
+        # csrf value length is too short
+        if len(csrf_value) < self.MIN_LENGTH:
+            return False
+
+        # csrf value entropy is too low, i.e. csrf value is too predictive
+        if self.entropy(csrf_value) < self.MIN_ENTROPY:
+            return False
+
+        return True
+
+    @staticmethod
+    def is_same_response(original_response: Page, mutated_response: Page):
+        """Check whether mutated_response status and content are identical to those of original_response"""
```

Comment:
I'm a bit concerned with that method but I think it is impossible to have perfect results for this module.

The way Wapiti crawls a website may result in a token being changed before the form is submitted if the website change the token after each request so there may be cases where the `original_response` informations aren't useful.

Reply:
> I'm a bit concerned with that method but I think it is impossible to have perfect results for this module.

This is a first step to compare the website behavior when sending valid and invalid csrf tokens, but i'm aware that comparing response contents might require to use something like a distance calculation, in order to take in account little variations (new csrf token value for instance). However it won't raise false positives as it is.

> The way Wapiti crawls a website may result in a token being changed before the form is submitted if the website change the token after each request so there may be cases where the original_response informations aren't useful.

Is this linked to #31 ?

Reply:
For the time being comparing the response should be enough.

Wapiti use a FIFO queue to crawl URLs discovered so if it crawls a web page with a form this one will be at the end of the queue and other webpages (those at the same "depth" that the current one) will be processed before the form.

It may have an impact on the token if the website is changing the token at every webpage fetched by the visitor... it depends on the way the website implement this